pyro/infer/mcmc/mh.py

```diff
@@ -0,0 +1,255 @@
+from __future__ import absolute_import, division, print_function
+
+import pyro
+import pyro.distributions as dist
+from pyro import poutine
+from pyro.infer.mcmc.trace_kernel import TraceKernel
+from pyro.util import ng_ones, ng_zeros
+from torch import Tensor as T
+from torch.autograd import Variable as V
+from numpy.random import choice
+from numpy import isfinite
+from torch import log as tlog
+
+_R = '_RETURN'
+
+
+def VT(val):
+    return V(T(val))
+
+
+def VTA(val):
+    return VT([val])
+
+
+def assert_valid_trace(trace):
+    assert not any(['fn' in n and isinstance(n['fn'], pyro._Subsample)
+                    for nid, n in trace.nodes(data=True)]), \
+                    "Single Site MH does not currently handle mapdata in model."
+
+
+# get all the sample sites
+def sample_sites(trace, minus=set()):
+    return [nid for nid, n in trace.nodes(data=True)
+            if nid not in minus and n["type"] == 'sample'
+            and not n["is_observed"]]
+
+
+class NormalProposal():
+    def __init__(self, mu, sigma, tune_frequency=100):
+        self.mu = mu
+        self.sigma = sigma
+        self.scale = 1.
+        self.tune_frequency = tune_frequency
+        self._tune_cnt = self.tune_frequency
+
+    def sample(self, site_node):
+        cur_val = site_node['value']
+        mu = self.mu.expand_as(cur_val)
+        sigma = self.scale*self.sigma
+
+        # sigma just a number? expand to equal the must
+        if tuple(sigma.data.shape) != mu.data.shape:
+            sigma = sigma.expand_as(mu.data)
+
+        return cur_val + dist.Normal(mu, sigma).sample()
+
+    def log_pdf_given(self, y, given_x):
+        # center our distribution at mu = given_x, then
+        # get log pdf of y
+        mu = given_x
+        sigma = self.scale*self.sigma
+        return dist.Normal(mu, sigma).log_pdf(y)
+
+    # borrowed from:
+    # https://github.com/mcleonard/sampyl/blob/master/sampyl/samplers/metropolis.py#L102
+    # which in turn borrows from pymc3
+    def tune(self, acceptance):
+        self._tune_cnt -= 1
+        if self._tune_cnt > 0:
+            return self
+
+        scale = self.scale
+        # Switch statement
+        if acceptance < 0.001:
+            # reduce by 90 percent
+            scale *= 0.1
+        elif acceptance < 0.05:
+            # reduce by 50 percent
+            scale *= 0.5
+        elif acceptance < 0.2:
+            # reduce by ten percent
+            scale *= 0.9
+        elif acceptance > 0.95:
+            # increase by factor of ten
+            scale *= 10.0
+        elif acceptance > 0.75:
+            # increase by double
+            scale *= 2.0
+        elif acceptance > 0.5:
+            # increase by ten percent
+            scale *= 1.1
+
+        self.scale = scale
+        # reset our tuning count
+        self._tune_cnt = self.tune_frequency
+        return self
+
+
+class MH(TraceKernel):
```

Comment:
Lets also add a docstring for this kernel class, and put in any relevant citations here, so as to have it reflect in our documentation.